### PR TITLE
Note highlight

### PR DESCRIPTION
**Pull request by jmattheis**

### CODE REVIEW

File: packages/snage/src/config/validator.ts [hunk 1]
```diff
@@ -6,17 +6,20 @@ import {currentSchema, currentVersion, getSchema} from './schema';
...
+    note: {links: []},
 };
 
 export const parseRawConfig = (yamlDoc: Document): E.Either<string, RawConfig> => {
     return pipe(
         migrateConfig(yamlDoc),
-        E.map((parsedConfig) => ({...fieldsWithDefaults, ...parsedConfig.toJSON()}))
+        E.map((parsedConfig) => {
+            const json = parsedConfig.toJSON();
+            return {...fieldsWithDefaults, ...json, template: {...fieldsWithDefaults.template, ...json.template, note: {...fieldsWithDefaults.note, ...json.note}}};
```
Comment (tvrg) [inline packages/snage/src/config/validator.ts]:
```suggestion
            return {...fieldsWithDefaults, ...json, template: {...fieldsWithDefaults.template, ...json.template}, note: {...fieldsWithDefaults.note, ...json.note}};
```

File: packages/snage/src/config/migrate.ts [hunk 2]
```diff
@@ -58,6 +58,12 @@ export const findIndex = (contents: YAMLMap, key: string): O.Option<number> => {
     return index === -1 ? O.none : O.some(index);
 };
 
+export const asYAMLMap = (obj: object): YAMLMap =>
+    Object.keys(obj).reduce((map, key) => {
+        map.set(key, obj[key]);
+        return map;
+    }, new YAMLMap());
+
```
Comment (tvrg) [inline packages/snage/src/config/migrate.ts:66]: We could probably also use [`YAML.createNode(obj)`](https://eemeli.org/yaml/#yaml-createnode). But we would need to cast the result from `Node` to `YAMLMap`.

```suggestion
export const asYAMLMap = (obj: Record<string, unknown>): YAMLMap => YAML.createNode(obj) as YAMLMap;
```

Comment (tvrg) [reply packages/snage/src/config/migrate.ts:66]: Well, no need to cast anything, if we directly use `YAML.createNode()` without wrapping it in another method I guess.

File: packages/snage/src/config/migration/v0-v2.test.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import {migrate} from '../migrate';
+import YAML from 'yaml';
+
+describe('v0 -> v2', () => {
```
Comment (tvrg) [inline packages/snage/src/config/migration/v0-v2.test.ts:4]: This error might appear in future migrations as well. What about making this test a bit more generic: Always migrate from `v0` to the current latest version?

File: packages/snage/src/config/migration/v1.test.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import {migrate} from '../migrate';
 import YAML from 'yaml';
 
-describe('v1 -> v2', () => {
+describe('v0 -> v1', () => {
```
Comment (tvrg) [inline packages/snage/src/config/migration/v1.test.ts:4]: woops, nice :+1:

File: changelog/104-style.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+type: change
+audience: user
+date: 2020-06-11
+components:
+  - config
+---
+# Add note styling
+
+You can define note styling inside your config.
+`css` will be applied to the note if the query `on` evaluates to `true`.
```
Comment (tvrg) [inline changelog/104-style.md:12]:
```suggestion
`css` will be applied to all notes for which the query `on` evaluates to `true`.
```

File: packages/shared/type.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ export interface ApiNote {
     id: string;
     content: string;
     summary: string;
+    style: Record<string, string>;
```
Comment (tvrg) [inline packages/shared/type.ts:5]: Is `string` valid for all css properties or do you have to supply numbers for some properties?

File: packages/snage/src/config/load.test.ts [hunk 7]
```diff
@@ -55,6 +57,43 @@ describe('parseConfig', () => {
         const links = conf.right.note.links({version: semver.parse('1.0.0')!, name: 'hello'});
         expect(links).toEqual([{label: 'Name hello 1.0.0', href: 'http://github.com'}]);
     });
+    it('creates css provider', () => {
+        const config: RawConfig = {
+            basedir: 'aeu',
```
Comment (tvrg) [inline packages/snage/src/config/load.test.ts:62]: That's the best basedir

File: packages/snage/src/config/load.ts [hunk 8]
```diff
@@ -93,18 +96,21 @@ const convert = (rawConfig: RawConfig, configFilePath: string): E.Either<string,
     return pipe(
         A.array.traverse(E.either)(rawConfig.fields, toField),
         E.chain((fields) =>
-            E.either.map(
-                createLinkProvider(fields, rawConfig.links),
-                (linkProvider): Config => ({
+            pipe(
+                sequenceS(E.either)({
```
Comment (tvrg) [inline packages/snage/src/config/load.ts:100]: awesome fp-ts is awesome :laughing:

File: packages/snage/src/config/load.ts [hunk 9]
```diff
@@ -128,6 +134,25 @@ const createSingleLinkProvider = (fields: Field[]) => (index: number, link: Link
...
 
+const createStyleProvider = (fields: Field[], styles: ConditionalStyle[]): E.Either<string, CSSProvider> => {
+    const parser = createParser(fields);
+
+    return pipe(
+        A.array.traverse(E.either)(styles, ({on, css}) =>
+            E.either.bimap(
+                parser(on),
+                (e) => `Invalid expression "${on}" ${JSON.stringify(e)}`,
+                (expression) => ({matcher: createMatcher(expression, fields), css} as const)
```
Comment (tvrg) [inline packages/snage/src/config/load.ts:145]: Dammit, we've forgotten to check out the `Optional`-variant for `CSSProvider`. Now we'll never know if it is better.

File: packages/snage/src/config/validator.ts [hunk 10]
```diff
@@ -10,15 +10,20 @@ const fieldsWithDefaults: Partial<RawConfig> = {
...
+    note: {links: [], styles: []},
 };
 
 export const parseRawConfig = (yamlDoc: Document): E.Either<string, RawConfig> => {
     return pipe(
         migrateConfig(yamlDoc),
         E.map((parsedConfig) => {
             const json = parsedConfig.toJSON();
-            return {...fieldsWithDefaults, ...json, template: {...fieldsWithDefaults.template, ...json.template, note: {...fieldsWithDefaults.note, ...json.note}}};
+            return {
```
Comment (tvrg) [inline packages/snage/src/config/validator.ts:21]: If this get's more complicated in the future, we should use lodash#merge :)

File: packages/snage/src/config/validator.ts (hunk 1 above)
Comment (tvrg) [reply packages/snage/src/config/validator.ts]: Is already resolved in a later commit

File: packages/shared/type.ts (hunk 6 above)
Comment (jmattheis) [reply packages/shared/type.ts:5]: Yeah, react allows numbers but everything can be configured via strings.

**Pull request merged**